### PR TITLE
applications: asset_tracker_v2: wifi location on AWS/Azure

### PR DESCRIPTION
**Pull request by maxd-nordic**
This patch combines location encoding (wifi+ncell) for the AWS and Azure backend.

Corresponding change on the cloud side: https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/issues/659

### CODE REVIEW

File: applications/asset_tracker_v2/src/cloud/aws_iot_integration.c [hunk 1]
```diff
@@ -28,8 +28,8 @@ LOG_MODULE_REGISTER(MODULE, CONFIG_CLOUD_INTEGRATION_LOG_LEVEL);
 #define BATCH_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 6)
 #define MESSAGES_TOPIC "%s/messages"
 #define MESSAGES_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 9)
-#define NEIGHBOR_CELLS_TOPIC "%s/ncellmeas"
-#define NEIGHBOR_CELLS_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 10)
+#define GROUNDFIX_TOPIC "%s/ground-fix"
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/aws_iot_integration.c:31]: `GROUND_FIX` is the most common naming convention used throughout NCS. I think we should separate the words in the ATv2 implementation as well to be consistent with the rest of the tree.

File: applications/asset_tracker_v2/src/cloud/cloud_codec/azure_iot_hub/json_protocol_names_azure_iot_hub.h [hunk 2]
```diff
@@ -59,6 +59,7 @@
 #define DATA_GNSS_HEADING   "hdg"
 #define DATA_GNSS_ACCURACY  "acc"
 
+#define DATA_NEIGHBOR_CELLS_ROOT "lte"
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/azure_iot_hub/json_protocol_names_azure_iot_hub.h:62]: Alignment

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 3]
```diff
@@ -643,23 +669,109 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+
+	cJSON *ap_array = cJSON_CreateArray();
+
+	if (ap_array == NULL) {
+		cJSON_Delete(root);
+		return err;
+	}
+
+	for (size_t i = 0; i < data->cnt; ++i) {
+		char str_buf[sizeof("112233445566")];
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:739]:
```suggestion
		char str_buf[AP_STRING_SIZE_MAX];
```
Lets avoid using magic numbers :magic_wand:

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 4]
```diff
@@ -643,23 +669,109 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+		struct wifi_scan_result const *const ap = (data->ap_info + i);
+
+		int ret;
+
+		/* MAC address is the only required parameter for the API call */
+		ret = snprintk(str_buf, sizeof(str_buf),
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:745]:
```suggestion
		int ret = snprintk(str_buf, sizeof(str_buf),
```

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 5]
```diff
@@ -643,23 +669,109 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+	if (ap_array == NULL) {
+		cJSON_Delete(root);
+		return err;
+	}
+
+	for (size_t i = 0; i < data->cnt; ++i) {
+		char str_buf[sizeof("112233445566")];
+		struct wifi_scan_result const *const ap = (data->ap_info + i);
+
+		int ret;
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:742]:
```suggestion
```

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 6]
```diff
@@ -643,23 +669,109 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+
+		if (!cJSON_AddItemToArray(ap_array, ap_obj)) {
+			cJSON_Delete(ap_obj);
+			err = -ENOMEM;
+			goto cleanup;
+		}
+	}
+
+	err = op_code_handle(root, JSON_COMMON_ADD_DATA_TO_OBJECT,
+				     DATA_WIFI_AP_MEAS, ap_array, NULL);
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:760]: Alignment

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c (hunk 4 above)
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:745]: Missing return value check

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 7]
```diff
@@ -643,23 +669,109 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+	if (err) {
+		goto cleanup;
+	}
 
+	json_add_obj(parent, DATA_WIFI_ROOT, root);
 	data->queued = false;
 	return err;
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:767]:
```suggestion
	return err;

```

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 8]
```diff
@@ -643,23 +669,109 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+cleanup:
+	cJSON_Delete(ap_array);
+	cJSON_Delete(root);
+	LOG_ERR("Failed to format WiFi location request, out of memory");
+	return err;
 }
+#endif
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:774]:
```suggestion
#endif /* CONFIG_LOCATION_METHOD_WIFI */
```

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.h [hunk 9]
```diff
@@ -198,6 +198,20 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
 					struct cloud_data_neighbor_cells *data,
 					enum json_common_op_code op);
 
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+/**
+ * @brief
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.h:203]: Missing docs

File: applications/asset_tracker_v2/src/cloud/aws_iot_integration.c (hunk 1 above)
Comment (maxd-nordic) [reply applications/asset_tracker_v2/src/cloud/aws_iot_integration.c:31]: fixed

File: applications/asset_tracker_v2/src/cloud/cloud_codec/azure_iot_hub/json_protocol_names_azure_iot_hub.h (hunk 2 above)
Comment (maxd-nordic) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/azure_iot_hub/json_protocol_names_azure_iot_hub.h:62]: fixed

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c (hunk 4 above)
Comment (maxd-nordic) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:745]: fixed

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c (hunk 6 above)
Comment (maxd-nordic) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:760]: fixed

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.h (hunk 9 above)
Comment (maxd-nordic) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.h:203]: fixed

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c (hunk 3 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:739]: However, if using a macro nothing is left to the imagination

Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:739]: `MAC_ADDRESS_STRING_SIZE_MAX` is probably better

**Comment by maxd-nordic** [comment]
@simensrostad thanks for the review! Could you have another look?

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c [hunk 10]
```diff
@@ -12,10 +12,13 @@
 #include "json_common.h"
 #include "json_helpers.h"
 #include "json_protocol_names.h"
+#include <net/nrf_cloud_location.h>
 
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(json_common, CONFIG_CLOUD_CODEC_LOG_LEVEL);
 
+#define AP_STRING_SIZE	sizeof("deadbeef1122")
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:20]:
```suggestion
/* MAC address string size including null character */
#define AP_STRING_SIZE	13
```

Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.c:20]: Why not just write the size directly? This way the user don't need to count the characters if their debugging this code for some reason and want to know.

File: applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.h [hunk 11]
```diff
@@ -198,6 +198,23 @@ int json_common_neighbor_cells_data_add(cJSON *parent,
...
+/**
+ * @brief Encode and add wifi survey data to the parent object.
+ *
+ * @param[out] parent Pointer to object that the encoded data is added to.
+ * @param[in] data Pointer to data that is to be encoded.
+ * @param[in] op Operation that is to be carried out.
+ *
+ * @retval 0 on success.
+ * @retval -ENODATA if the passed in data is not queued or has an invalid timestamp value.
+ * @retval -ENOMEM if the function fails to allocate memory.
```
Comment (simensrostad) [inline applications/asset_tracker_v2/src/cloud/cloud_codec/json_common.h:211]: Missing for -EFAULT

File: applications/asset_tracker_v2/tests/json_common/src/main.c [hunk 12]
```diff
@@ -37,6 +37,7 @@ static int encoded_output_check(cJSON *object, char *validation_string, int8_t q
 
 	if (strcmp(validation_string, dummy.buffer) != 0) {
 		/* Dummy buffer should equal the validation string. */
+		printf("validation: %s, buffer: %s", validation_string, dummy.buffer);
```
Comment (simensrostad) [inline applications/asset_tracker_v2/tests/json_common/src/main.c:40]:
```suggestion
		printk("validation: %s, buffer: %s", validation_string, dummy.buffer);
```

File: applications/asset_tracker_v2/tests/json_common/src/main.c [hunk 13]
```diff
@@ -1244,6 +1287,11 @@ void test_main(void)
 					       test_setup_object,
 					       test_teardown_object),
 
+		/* Neighbor cell */
```
Comment (simensrostad) [inline applications/asset_tracker_v2/tests/json_common/src/main.c:1290]: Copy paste shananigans?

**Review by simensrostad — APPROVED** [review]

**Review by coderbyheart — APPROVED** [review]

**Pull request merged**